src/util/parallel/data.ml

```diff
@@ -0,0 +1,219 @@
+          let new_bucket = Bucket.create key in
+          let success = Atomic.compare_and_set bucket None (Some new_bucket) in
+          if success then (
+            (new_bucket.value, true) 
+          )
+          else find_create_inner hm key hash buckets 
+        | Some bucket -> 
+          Bucket.find_create bucket key
+      in
+      let current_generation = Atomic.get hm.resize_generation in
+      let hash = abs @@ H.hash key in
+      let value, was_created = find_create_inner hm key hash (Atomic.get hm.buckets) in
+      if (current_generation mod 2 = 0) && (Atomic.get hm.resize_generation = current_generation || not was_created) then (
+        if (Atomic.get hm.nr_elements >= Atomic.get hm.nr_buckets * 2) then (
+          resize hm;
+        );
+        if (was_created) then Atomic.incr hm.nr_elements;
+        (value, was_created)
+      )
+      else (
+        while (Atomic.get hm.resize_generation = current_generation) do () (* spin *)
+        done;
+        find_create hm key)
+
+
+    and resize hm =
+      let current_generation = Atomic.get hm.resize_generation in
+      if ((current_generation mod 2 = 0) && Atomic.compare_and_set hm.resize_generation current_generation (current_generation+1)) then (
+
+        let old_size = Atomic.get hm.nr_buckets in
+        let new_size = old_size * 2 in
+
+        (* Note that we need a new atomic for each element, so we need Array.init *)
+        let new_buckets = Array.init new_size (fun _ -> Atomic.make None) in
+        let rec rehash_bucket (bucket: Bucket.t option Atomic.t) =
+          match Atomic.get bucket with
+          | None -> ()
+          | Some {key; value; next} -> 
+            begin
+              let hash = abs @@ H.hash key in
+              let new_location = Array.get new_buckets (hash mod new_size) in
+              let _ = match Atomic.get new_location with
+                | None -> (ignore @@ Atomic.set new_location (Some (Bucket.create_with_value key value)))
+                | Some new_bucket -> 
+                  let newer_bucket = Bucket.create_with_value_and_next key value new_bucket in
+                  (ignore @@ Atomic.set new_location (Some newer_bucket)) in
+              rehash_bucket next
+            end
+        in
+        Array.iter rehash_bucket (Atomic.get hm.buckets);
+
+        Atomic.set hm.buckets new_buckets;
+        Atomic.set hm.nr_buckets new_size;
+        Atomic.incr hm.resize_generation;
+      )
+
+    let to_value_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) in 
+      let rec bucket_to_value_seq (bucket : Bucket.t option Atomic.t) = match Atomic.get bucket with
+        | None -> Seq.empty
+        | Some b -> fun () -> Seq.Cons (b.value, bucket_to_value_seq b.next) in
+      Seq.flat_map bucket_to_value_seq bucket_seq
+
+    let to_hashtbl hm =
+      let ht = HM.create 10 in
+      let seq = to_seq hm in
+      Seq.iter (fun (k, v) -> HM.add ht k (Atomic.get v)) seq;
+      ht
```

Comment:
Isn't this like a `Hastbl.of_seq` after a `Seq.map`?

Reply:
Batteries hashtable has no of_seq.